MCprep_addon/optimize_scene.py

```diff
@@ -218,10 +237,63 @@ def execute(self, context):
 		for mat in bpy.data.materials:
 			matGen = util.nameGeneralize(mat.name)
 			canon, form = generate.get_mc_canonical_name(matGen)
+			mat_type = None
+			homogenous_volumes = 0
+			not_homogenous_volumes = 0
 			if generate.checklist(canon, "reflective"):
 				Glossy += 1
+				mat_type = "reflective"
 			if generate.checklist(canon, "glass"):
 				Transmissive += 1
+				mat_type = "glass"
+
+			if mat.use_nodes:
+				nodes = mat.node_tree.nodes
+				for node in nodes:
+					if node.bl_idname == "ShaderNodeVolumeScatter" or \
+						node.bl_idname == "ShaderNodeVolumeAbsorption" or \
+						node.bl_idname == "ShaderNodeVolumePrincipled":
+						density_socket = node.inputs["Density"] # Grab the density
```

Comment:
Let's move this indented level and below into another function if possible, it's getting way too indented here.